### PR TITLE
window.c: fix floating window border width calculation

### LINKED ISSUE #14478: src/nvim/ui_compositor.c:369: compose_line: Assertion `until <= default_grid.Columns` failed
## environment
- `nvim --version`: latest master (b227cedf82341813514da55baa48511eca4014b0)
- Operating system/version: ubuntu 20.04
- reproduce in tui and nvim-qt

## Steps to reproduce
This bug is reliably reproducible using LSP hover.
I believe it's possible to reproduce it only using `nvim_open_win()`,
but I couldn't find the right combination of arguments.

0. Preparation
    * `example.lua` 
        ```lua
        require'asdaf'
        ```
    * `mini.vim`
        ```vim
        set runtimepath^=/path/to/nvim-lspconfig

        " foldcolumn=2 also works. foldcolumn=1 doesn't.
        set number
        " make the window small
        set lines=8 columns=80

        lua << EOF
        require'lspconfig'.sumneko_lua.setup {
          -- your configuration
        }
        EOF
        ```

1. `nvim -u mini.vim example.lua`
2. `<C-w>v<C-w><C-l>` (right vsplit)
3. `lua vim.lsp.buf.hover()` on "require"
4. `<C-p>` (move the cursor into the floating winow)
5. `<C-d>`

```
nvim: ../src/nvim/ui_compositor.c:369: compose_line: Assertion `until <= default_grid.Columns' failed
Aborted (core dumped)
```

## note
* related? https://github.com/neovim/neovim/issues/9989

### PR DESCRIPTION
**Pull request by mjlbach**
Closes #14478 and #14377 
* lift calculation of w_border_adj into win_config_float
* Check max floating window width agains w_width_outer instead of
  w_width when position window

### CODE REVIEW

**Comment by bfredl** [comment]
cool, the height calculations can be adjusted analogically as well :)

**Comment by mjlbach** [comment]
Better?

File: src/nvim/window.c [hunk 1]
```diff
@@ -676,9 +676,17 @@ void win_config_float(win_T *wp, FloatConfig fconfig)
...
+  for (int i = 0; i < 4; i++) {
+    wp->w_border_adj[i] =
+      has_border && wp->w_float_config.border_chars[2 * i+1][0];
+  }
+
   if (!ui_has(kUIMultigrid)) {
-    wp->w_height = MIN(wp->w_height, Rows-1);
-    wp->w_width = MIN(wp->w_width, Columns);
+    wp->w_height = MIN(wp->w_height, 
+		       Rows - (1 + wp->w_border_adj[0] + wp->w_border_adj[2]));
```
Comment (bfredl) [inline src/nvim/window.c:687]: I would have `Rows - 1 - (wp->w_border_adj[0] + wp->w_border_adj[2])` as the -1 is unrelated.

Comment (mjlbach) [reply src/nvim/window.c]: Fixed

**Comment by bfredl** [comment]
perhaps a test? :)

**Comment by mjlbach** [comment]
I haven't been able to reproduce the crash via tests. I tried various widths to be sure I believe the issue should occur with col 5/width 35, maybe the text content isn't enough as I recall it had originally occured only when scrolling down in the floating window:

```lua
    it('floating windows with border can terminate inner extent on edge of grid', function()
      screen:try_resize(40,2)
      local buf = meths.create_buf(false, false)
      meths.buf_set_lines(buf, 0, -1, true, {' halloj! ',
                                             ' BORDAA  '})
      local win = meths.open_win(buf, false, {relative='editor', width=34, height=2, row=2, col=5, border="single"})
      feed '<c-w><c-w>'
      feed '<c-d><c-d><c-d>'
    end)
```

**Comment by bfredl** [comment]
well, it is good enough for me a test that shows that it _looks_ right. i e the float with borders occupy the screen correctly, even when the configured size is over-specified.

**Comment by mjlbach** [comment]
I think it would be a good idea to find a test that reproduces the original problem since we already have many border tests, I guess the main difference would be a border test where the floating window spans the exact terminal width

**Comment by bfredl** [comment]
The original problem IS that the float gets incorrectly sized/positioned and thus incorrectly drawn. drawing outside the window might or might not cause a crash depending on various circumstances, but it will always look horrendously incorrect. Showing that the float is correctly drawn is enough to demonstrate that nvim is not attempting to draw outside the screen.

**Comment by mjlbach** [comment]
@bfredl tests are all green, let me know if you want any changes to anything :)

**Comment by bfredl** [comment]
seems all good :)

**Pull request merged**